projects/project-2/assignment/Assginment 2.md

```diff
@@ -0,0 +1,241 @@
+
+<pre>
+  <code>
+PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
+PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
+PREFIX owl: <http://www.w3.org/2002/07/owl#>
+
+SELECT ?property ?class ?restriction
+WHERE {
+    # Find properties with universal restrictions
+    ?property a owl:ObjectProperty .
+    ?restriction a owl:Restriction ;
+                 owl:onProperty ?property ;
+                 owl:allValuesFrom ?class .
+    
+    # Check for absence of existential restrictions on the same property
+    FILTER NOT EXISTS {
+        ?property a owl:ObjectProperty .
+        ?existentialRestriction a owl:Restriction ;
+                                owl:onProperty ?property ;
+                                owl:someValuesFrom ?class .
+    }
+}
+    
+  </code>
+</pre>
+
+
+4. Misusing primitive and defined classes: detecting the failure to make the definition ‘complete’ rather than ‘partial.’ In other words,
+   finding definitions that are ‘necessary and sufficient’ rather than just ‘necessary'. (Level 3 - 20 pts)
+
+<pre>
+  <code>
+    PREFIX : <http://example.org/ontology/>
+
+SELECT ?definition ?necessaryCondition
+WHERE {
+    ?definition a :Definition ;
+                :necessaryCondition ?necessaryCondition .
+    
+    FILTER NOT EXISTS {
+        ?definition :sufficientCondition ?sufficientCondition .
+    }
+}
+
+  </code>
+</pre>
+
+
+5. detects definitions of relationships that are defined as inverse of another relationship when the two relationships are `owl:SymmetricProperty`.
+   For example, detecting the creation of the symmetric relationship “farFrom” an inverse relationship, e.g. itself, “farFrom”. (Level 5 - 5 pts)
+
+<pre>
+  <code>
+PREFIX owl: <http://www.w3.org/2002/07/owl#>
+PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
+PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
+
+SELECT ?property ?inverseProperty
+WHERE {
+    # Find properties that are declared symmetric
+    ?property a owl:SymmetricProperty .
+    
+    # Check if the property is defined as the inverse of itself or another property
+    ?property owl:inverseOf ?inverseProperty .
+    
+    # Check if the inverse property is the same as the symmetric property itself or also symmetric
+    FILTER (?property = ?inverseProperty || EXISTS { ?inverseProperty a owl:SymmetricProperty })
+}
+    
+  </code>
+</pre>
+
+6. Detect Classes Without a Subclass (Leaf Classes). This query detects classes that do not have any subclasses.
+   These "leaf" classes might be indicators of incomplete modeling if too many are present. (level 3 - 20 pts)
+
+<pre>
+  <code>
+PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
+PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
+
+SELECT ?class
+WHERE {
+    ?class a rdfs:Class .
+    
+    # Find classes that do not have any subclasses
+    FILTER NOT EXISTS {
+        ?subClass rdfs:subClassOf ?class .
+    }
```

Comment:
I don't think having a lot of leaf classes is a problem. But if that's what you want to do, you should probably add a COUNT function to see how many there are

Reply:
Okay! Thank you!